File: src/main/java/alfheim/common/item/block/ItemBlockAnomaly.java
package alfheim.common.item.block;

import static vazkii.botania.common.core.helper.ItemNBTHelper.*;

import java.util.HashMap;

import alfheim.api.AlfheimAPI;
import alfheim.api.ModInfo;
import alfheim.api.block.tile.SubTileEntity;
import alfheim.common.block.BlockAnomaly;
import alfheim.common.block.tile.TileAnomaly;
import alfheim.common.core.registry.AlfheimBlocks;
import cpw.mods.fml.relauncher.Side;
import cpw.mods.fml.relauncher.SideOnly;
import net.minecraft.block.Block;
import net.minecraft.client.renderer.texture.IIconRegister;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.ItemBlock;
import net.minecraft.item.ItemStack;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.IIcon;
import net.minecraft.world.World;
import vazkii.botania.common.core.helper.ItemNBTHelper;

public class ItemBlockAnomaly extends ItemBlock {
	
	public static final String TYPE_UNDEFINED = "undefined";
	
	public ItemBlockAnomaly(Block block) {
		super(block);
	}
	
	@Override
	public IIcon getIcon(ItemStack stack, int pass) {
		return BlockAnomaly.iconUndefined;
	}
	
	@Override
	public String getUnlocalizedName(ItemStack stack) {
		return "tile.Anomaly." + ItemNBTHelper.getString(stack, SubTileEntity.TAG_TYPE, TYPE_UNDEFINED);
	}
	
	@Override
	public int getMetadata(int meta) {
		return meta;
	}
	
	public static String getType(ItemStack stack) {
		return ItemNBTHelper.detectNBT(stack) ? ItemNBTHelper.getString(stack, SubTileEntity.TAG_TYPE, TYPE_UNDEFINED) : TYPE_UNDEFINED;
	}
	
	public static ItemStack ofType(String type) {
		return ofType(new ItemStack(AlfheimBlocks.anomaly), type);
	}
	
	public static ItemStack ofType(ItemStack stack, String type) {
		if (type == null || type.isEmpty()) type = TYPE_UNDEFINED;
		ItemNBTHelper.setString(stack, SubTileEntity.TAG_TYPE, type);
		return stack;
	}
	
	@Override
	public boolean placeBlockAt(ItemStack stack, EntityPlayer player, World world, int x, int y, int z, int side, float hitX, float hitY, float hitZ, int metadata) {
		boolean placed = super.placeBlockAt(stack, player, world, x, y, z, side, hitX, hitY, hitZ, metadata);
		if(placed) {
			String type = getType(stack);
			TileEntity te = world.getTileEntity(x, y, z);
			if(te instanceof TileAnomaly) {
				/*TileAnomaly tile = */
				((TileAnomaly) te).addSubTile(type);
//				tile.onBlockAdded(world, x, y, z);
//				tile.onBlockPlacedBy(world, x, y, z, player, stack);
				if(!world.isRemote)
					world.markBlockForUpdate(x, y, z);
			}
		}
		
		return placed;
	}
}

File: src/main/java/alfheim/common/item/block/ItemBlockAnomaly.kt
package alfheim.common.item.block

import alexsocol.asjlib.ASJUtilities
import alfheim.api.ModInfo
import alfheim.common.block.*
import alfheim.common.block.tile.TileAnomaly
import alfheim.common.block.tile.TileAnomaly.Companion.TAG_SUBTILE_COUNT
import alfheim.common.block.tile.TileAnomaly.Companion.TAG_SUBTILE_MAIN
import alfheim.common.block.tile.TileAnomaly.Companion.TAG_SUBTILE_NAME
import net.minecraft.block.Block
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.*
import net.minecraft.util.IIcon
import net.minecraft.world.World
import vazkii.botania.api.lexicon.IRecipeKeyProvider
import vazkii.botania.common.core.helper.ItemNBTHelper.*

class ItemBlockAnomaly(block: Block): ItemBlock(block), IRecipeKeyProvider {
	
	init {
		maxStackSize = 1
		setTextureName(ModInfo.MODID + ":undefined")
	}
	
	override fun getIcon(stack: ItemStack, pass: Int): IIcon {
		return BlockAnomaly.iconUndefined
	}
	
	override fun getUnlocalizedName(stack: ItemStack?): String {
		return "tile.Anomaly." + getString(stack, TAG_SUBTILE_MAIN, TYPE_UNDEFINED)
	}
	
	override fun getMetadata(meta: Int): Int {
		return meta
	}
	
	override fun placeBlockAt(stack: ItemStack, player: EntityPlayer, world: World, x: Int, y: Int, z: Int, side: Int, hitX: Float, hitY: Float, hitZ: Float, metadata: Int): Boolean {
		if (!super.placeBlockAt(stack, player, world, x, y, z, side, hitX, hitY, hitZ, metadata)) return false
		
		if (!player.capabilities.isCreativeMode) return true
		
		val te = world.getTileEntity(x, y, z) as? TileAnomaly ?: return true
		te.readCustomNBT(getNBT(stack))
		te.lock(x, y, z, world.provider.dimensionId)
		
		if (world.isRemote)
			return true
		
		world.markBlockForUpdate(x, y, z)
		ASJUtilities.dispatchTEToNearbyPlayers(te)
		
		return true
	}
	
	override fun getKey(stack: ItemStack) = "${stack.unlocalizedName}~${getType(stack)}"
	
	companion object {
		
		const val TYPE_UNDEFINED = "undefined"
		
		fun getType(stack: ItemStack): String {
			return if (detectNBT(stack)) getString(stack, TAG_SUBTILE_MAIN, TYPE_UNDEFINED) else TYPE_UNDEFINED
		}
		
		fun ofType(type: String): ItemStack {
			return ofType(ItemStack(AlfheimBlocks.anomaly), type)
		}
		
		fun ofType(stack: ItemStack, type: String?): ItemStack {
			var t = type
			if (type.isNullOrEmpty()) t = TYPE_UNDEFINED
			setString(stack, TAG_SUBTILE_MAIN, t)
			setInt(stack, TAG_SUBTILE_COUNT, 1)
			setString(stack, TAG_SUBTILE_NAME + "1", t)
			
			return stack
		}
	}
}
